### PR TITLE
lib/posix-poll: Autoremove closed files from epoll

### PR DESCRIPTION
**Pull request by andreittr**
### Description of changes

This change makes epoll automatically unregister any monitored files when the last reference to them is released, making use of ukfile finalizers.
Furthermore, to enable this change, vfscore vs ukfile shimming in epoll code has also been cleaned up a bit.

This behavior helps mitigate a design oversight of `epoll` where if applications close file descriptors before removing them from epoll's monitored set, they lose the ability to deregister said file without closing the entire epoll fd.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.

### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]

### Additional configuration
N/A

### CODE REVIEW

**Review by DeliaPavel — APPROVED** [review]
Looks good to me!
Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

File: lib/posix-poll/epoll.c [hunk 1]
```diff
@@ -182,16 +199,42 @@ static struct epoll_entry **epoll_find(const struct uk_file *epf, int fd,
...
+	struct epoll_entry **entp;
+
+#if CONFIG_LIBVFSCORE
+	UK_ASSERT(!ent->legacy);
+#endif /* CONFIG_LIBVFSCORE */
+
+	sf.file = ent->f;
+	uk_file_wlock(ent->epf);
+
+	uk_pr_err("Autodel %p\n", arg);
```
Comment (andreistan26) [inline lib/posix-poll/epoll.c]: Why is this an error?

**Review by andreistan26 — COMMENTED** [review]
Other than that it works great!

File: lib/posix-poll/epoll.c (hunk 1 above)
Comment (andreittr) [reply lib/posix-poll/epoll.c]: Oh my, good catch. It's mistakenly left over from development, will fix (remove or turn to info).

**Comment by andreittr** [comment]
PR ON HOLD pending fixing an unforeseen bug re: autoremoving lwip sockets. Turned to draft.

**Comment by andreittr** [comment]
PR reopened; bug fixed & comments addressed. Reviews re-requested.

**Review by andreistan26 — APPROVED** [review]
Looks good!

Reviewed-by: Andrei Stan [andreistan2003@gmail.com](mailto:andreistan2003@gmail.com)

**Review by DeliaPavel — APPROVED** [review]
Reviewed-by: Delia Pavel <delia_maria.pavel@stud.acs.upb.ro>

**Review by eduardvintila — APPROVED** [review]
All good, nice job.

Reviewed-by: Eduard Vintilă <eduard.vintila47@gmail.com>

**Review by razvand — APPROVED** [review]
Approved-by: Razvan Deaconescu <razvand@unikraft.io>

**Pull request merged**